pulsar/transaction_coordinator_client.go

```diff
@@ -20,23 +20,322 @@ package pulsar
+			return
+		case req := <-t.requestCh:
+			switch r := req.(type) {
+			case *newTxnOp:
+				t.newTransaction(r)
+			case *addPublishPartitionOp:
+				t.addPublishPartitionToTxn(r)
+			case *addSubscriptionOp:
+				t.addSubscriptionToTxn(r)
+			case *endTxnOp:
+				t.endTxn(r)
+			}
+		case <-t.connectClosedCh:
+			t.log.Infof("Transaction handler %d will reconnect to the transaction coordinator", t.partition)
+			t.reconnectToBroker()
+		}
+	}
+}
+
+func (t *transactionHandler) reconnectToBroker() {
+	var delayReconnectTime time.Duration
+	var defaultBackoff = internal.DefaultBackoff{}
+
+	for {
+		if t.getState() == txnHandlerClosed {
+			// The handler is already closing
+			t.log.Info("transaction handler is closed, exit reconnect")
+			return
+		}
+
+		delayReconnectTime = defaultBackoff.Next()
+
+		t.log.WithFields(log.Fields{
+			"delayReconnectTime": delayReconnectTime,
+		}).Info("Reconnecting to broker")
+		time.Sleep(delayReconnectTime)
+
+		// double check
+		if t.getState() == txnHandlerClosed {
+			// Txn handler is already closing
+			t.log.Info("transaction handler is closed, exit reconnect")
+			return
+		}
+
+		err := t.grabConn()
+		if err == nil {
+			// Successfully reconnected
+			t.log.Info("Reconnected transaction handler to broker")
+			return
+		}
+		t.log.WithError(err).Error("Failed to create transaction handler at reconnect")
+		errMsg := err.Error()
+		if strings.Contains(errMsg, errMsgTopicNotFound) {
+			// when topic is deleted, we should give up reconnection.
+			t.log.Warn("Topic Not Found.")
+			break
+		}
+	}
+}
+
+func (t *transactionHandler) checkRetriableError(err error, op any) bool {
+	if err != nil && errors.Is(err, internal.ErrConnectionClosed) {
+		// We are in the EventLoop here, so we need to insert the request back to the requestCh asynchronously.
+		go func() {
+			t.requestCh <- op
+		}()
+		return true
+	}
+	return false
+}
+
+type newTxnOp struct {
+	// Request
+	timeout time.Duration
+
+	// Response
+	done  chan any
+	err   error
+	txnID *TxnID
+}
+
+func (t *transactionHandler) newTransaction(op *newTxnOp) {
+	requestID := t.tc.client.rpcClient.NewRequestID()
+	nextTcID := t.tc.nextTCNumber()
+	cmdNewTxn := &pb.CommandNewTxn{
+		RequestId:     proto.Uint64(requestID),
+		TcId:          proto.Uint64(nextTcID),
+		TxnTtlSeconds: proto.Uint64(uint64(op.timeout.Milliseconds())),
+	}
+	res, err := t.tc.client.rpcClient.RequestOnCnx(t.getConn(), requestID, pb.BaseCommand_NEW_TXN, cmdNewTxn)
+	if t.checkRetriableError(err, op) {
+		return
+	}
+	defer close(op.done)
+	if err != nil {
+		op.err = err
+	} else if res.Response.NewTxnResponse.Error != nil {
+		op.err = getErrorFromServerError(res.Response.NewTxnResponse.Error)
+	} else {
+		op.txnID = &TxnID{*res.Response.NewTxnResponse.TxnidMostBits,
+			*res.Response.NewTxnResponse.TxnidLeastBits}
+	}
+}
+
+type addPublishPartitionOp struct {
+	// Request
+	id         *TxnID
+	partitions []string
+
+	// Response
+	done chan any
+	err  error
```

Comment:
Why not use `chan error` as the response? Then you can replace `close(op.done)` with 

```go
	if err != nil {
		op.errorCh <- err
	} else if res.Response.AddPartitionToTxnResponse.Error != nil {
		op.errorCh <- getErrorFromServerError(res.Response.AddPartitionToTxnResponse.Error)
	} else {
		op.errorCh <- nil
	}
```

and you just need to return the value from the channel:

```go
return <-op.errorCh 
```

The same question is applied for other `xxxOp` structs.